site/content/blog/2021-11-17-database-sql.md

```diff
@@ -0,0 +1,219 @@
+
+This has served us well for several years, but as plugins started to be more powerful, they needed to maintain their own state in separate database tables. Initially, we just had a single table offering key-value semantics for all plugins via the plugin API, but clearly that wasn't enough. Plugins needed access to the database.
+
+This created a new challenge. Since each plugin was its own isolated process, each of them would open additional connections to the DB. And the more plugins we had, the more amplified this would become. Our customers started to complain that the system exceeded the `max_connections` config setting on their DB.
+
+![image](/blog/2021-11-17-database-sql/1.png)
+
+What we needed was to route all queries via the server so that a single connection pool would be used by all plugins. An obvious solution here is to put a connection proxy (like `PgBouncer` or `ProxySQL`) in front of the database to multiplex between connections. While that works (and indeed we use that in our Cloud environments), ours is primarily an on-prem software, which means we just cannot ask all of our customers to add a DB proxy in their stack.
+
+![image](/blog/2021-11-17-database-sql/2.png)
+
+### Approaches
+
+**1. Implement RPC APIs for the queries**
+
+The most straightforward solution would be to have RPC APIs for these DB calls and have the plugins make the RPC calls to the server, instead of using the DB directly.
+
+However, all of the queries contained logic that was relevant to the plugin, not the server. So it would not make sense to place that code in the server. And moreover, even if we did have a way to do that, writing API wrappers for every new DB call wasn't going to scale. SQL queries change very often and it would be a nightmare to maintain the RPC API. We needed something low-level.
+
+**2. DB API on top of RPC**
+
+A better alternative was to use a database query API, and have the API marshal the query params via RPC and execute them on the server. This was a generic solution and did not require you to write new APIs for every new query.
+
+However, we have different teams working on different plugins. A new query API would mean all of the existing code would need to be migrated to the new API, which a lot of teams were reluctant to do.
+
+What we needed was for the existing code to work without any change, but at the same time route the queries via RPC to the server instead of hitting the DB.
+
+**3. database/sql/driver**
+
+The `database/sql` package is just a set of methods that run the `database/sql/driver` interface. This driver interface is actually the one that communicates with the DB. Therefore, if we could implement our own driver interface that just routed all queries, then we could solve this problem with practically no change at all!
```

Comment:
Shall we add a couple of sentences about implicit interfaces in Go? It doesn't specifically solve the problem here but it's a very neat future when you are using interfaces.

Reply:
Hmm .. what do you mean by implicit interfaces?

Reply:
Uh, sorry. Some programming languages like Java requires you to import the interface explicitly and add keywords like `implements` whereas Go doesn't. You can ignore this comment if it doesn't make sense to you :)

Reply:
Ahh I see! Yeah, I think it would be taking a detour and going to explain language features. Keeping it short and tidy.